File: Difficulty: Medium/Undirected Graph Cycle/undirected-graph-cycle.cpp
class Solution {
  public:
    bool detect(int src, vector<int> adj[], vector<int>& vis){
        queue<pair<int, int>> q;
        q.push({src, -1});
        vis[src] = 1;
        
        while (!q.empty()) {
            int node = q.front().first;
            int parent = q.front().second;
            q.pop();
            
            for (auto adjNode : adj[node]) {
                if (!vis[adjNode]) {
                    vis[adjNode] = 1;
                    q.push({adjNode, node});
                } else if (adjNode != parent) {
                    return true;
                }
            }
        }
        return false;
    }
    bool isCycle(int V, vector<vector<int>>& edges) {
        // Code here
        vector<int> adj[V];
        for (auto edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        vector<int> vis(V,0);
        for(int i=0;i<V;i++){
            if(!vis[i]){
                if(detect(i,adj,vis)) return true;
        }
        }
        return false;
    }
};

File: Difficulty: Medium/Undirected Graph Cycle/undirected-graph-cycle.py
from collections import deque
class Solution:
	def isCycle(self, V, edges):
		#Code here
		adj = [[] for _ in range(V)]
		for u,v in edges:
		    adj[u].append(v)
		    adj[v].append(u)
		vis = [0]*V  
		def bfs_cycle(start):
            q = deque([(start, -1)])
            vis[start] = True
            
            while q:
                node, parent = q.popleft()
                for neighbor in adj[node]:
                    if not vis[neighbor]:
                        vis[neighbor] = True
                        q.append((neighbor, node))
                    elif neighbor != parent:
                        return True
            return False
        
        # Check all components
        for i in range(V):
            if not vis[i]:
                if bfs_cycle(i):
                    return True
        return False
		          
		        